### PR TITLE
Warn in colorbar() when mappable.axes != figure.gca().

### PR DESCRIPTION
**Pull request by anntzer**
## PR Summary

See discussion at https://github.com/matplotlib/matplotlib/pull/12333#issuecomment-425660082.  Already :+1:'d by @tacaswell :)

Edit: also closes https://github.com/matplotlib/matplotlib/issues/15986 indirectly.

Edit: will be rebased over https://github.com/matplotlib/matplotlib/pull/16058 once that goes in.

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/contour.py [hunk 1]
```diff
@@ -997,16 +997,21 @@ def __init__(self, ax, *args,
             warnings.warn('The following kwargs were not used by contour: ' +
                           s)
 
+    @cbook.deprecated("3.1")
```
Comment (jklymak) [inline lib/matplotlib/contour.py]: Did we decide to deprecate `self.ax`?

**Comment by jklymak** [comment]
I'm -0.5 on this change.    The implicit API is a bit goofy anyway, but there *are* plenty of reasons to have the colorbar steal from the current axes rather than the mappable's axes.  I guess I think its more unexpected for 
```python
plt.subplots(1, 2, 1)
pcm = plt.pcolormesh(X)
plt.subplots(1, 2, 2)
plt.plot(y)
plt.colorbar(pcm) 
```
to put the colorbar next to the first subplot than next to the one I specified last.

**Comment by anntzer** [comment]
For the change to colorbar() to work we need ScalarMappables to have an `.axes`, so ContourSets need to grow such an attribute (... or we need to special-case ContourSets in the colorbar() code, which seems worse).  We don't *have* to deprecate `.ax`, we can leave both of them around, but it seems just as good to make things consistent.
I can undeprecate `ContourSet.ax` if that's the main issue...

While `.ax` and `.axes` may be nearly as common in raw counts in the codebase, I'd guess `.axes` is much more common in actual use because that's what Artists have, and Artists also share a lot of code in the base class, whereas `.ax` is on a couple of custom non-Artist things (e.g. ContourSets) with each their own codepaths.

As for
```
plt.subplots(1, 2, 1)
pcm = plt.pcolormesh(X)
plt.subplots(1, 2, 2)
plt.plot(y)
plt.colorbar(pcm) 
```
if you make your plots vertically (211 and 212 instead of 121 and 122), I think the new behavior is better...

TBH I don't really care about `plt.colorbar()`, but at least for `fig.colorbar()` I'd rather not invoke the notion of "current axes".

**Comment by jklymak** [comment]
> TBH I don't really care about plt.colorbar(), but at least for fig.colorbar() I'd rather not invoke the notion of "current axes".

OK, I agree with that.  

I don't object to `axes` being universal versus `ax`, but I wasn't aware we had decided that.  And as noted above an effort should be made to do it everywhere.   Understood about needing the attribute.

**Comment by anntzer** [comment]
We haven't decided it, it's just needed for this specific PR to work (and again I can undeprecate `.ax` for this specific PR if you prefer).  Making the change everywhere is absolutely not the point of this PR.

**Comment by jklymak** [comment]
:+1:, but we should decide on the deprecation...

**Comment by timhoffm** [comment]
> For the change to colorbar() to work we need ScalarMappables to have an `.axes`, so ContourSets need to grow such an attribute.

Thanks for the explanation. Having `ax` and `axes` is not good either. Even though it's not the target of this PR, can we decide on a global change from `ax` to `axes` in attribute names? That would allow to keep the deprecation in here and add the remaining ones in another PR.

**Comment by anntzer** [comment]
I'd be happy with the change but I'm sure someone will complain about useless code churn, etc :)

**Comment by timhoffm** [comment]
@anntzer To move this forward, you could split out the `ax` -> `axes` deprecation and move that to #16058.

**Comment by anntzer** [comment]
OK, I moved the relevant changes to #16058 and will rebase this one on top of it once it goes in.

**Comment by anntzer** [comment]
rebased

File: doc/api/api_changes_3.4/deprecations.rst [hunk 2]
```diff
@@ -1,6 +1,20 @@
...
+
+Currently, `.Figure.colorbar` steals space by default from the current axes
+to place the colorbar.  In a future version, it will steal space from the
+mappable's axes instead.  In preparation for this change, `.Figure.colorbar`
+now emits a warning when the current axes is not the same as the mappable's
+axes.
+
+The ``.ax`` attribute to `.ContourSet` is deprecated in favor of
+``.axes`` for consistency with artists (even though `.ContourSet` is not an
+`.Artist` subclass).
```
Comment (timhoffm) [inline doc/api/api_changes_3.4/deprecations.rst:15]: Can be removed because it's been extracted to #16058.

Comment (anntzer) [reply doc/api/api_changes_3.4/deprecations.rst:15]: yup

**Comment by anntzer** [comment]
Actually, looking at this again, do we want to have the same behavior for pyplot.colorbar too?  Obviously pyplot.colorbar gets to know what the current image (gci()) is, but it likely makes sense to always use gci()'s axes, not the current axes?  (for the same reason why this makes sense for the OO API)

**Comment by anntzer** [comment]
Added the warning for pyplot.colorbar as a separate commit.

**Pull request merged**